rfcs/2020-08-21-3092-apache-metrics-source.md

```diff
@@ -0,0 +1,235 @@
+# RFC 3092 - 2020-08-21 - Apache HTTP Server metrics source
+
+This RFC is to introduce a new metrics source to consume metrics from the
+[Apache HTTP Server] (httpd). The high level plan is to implement a scrapper
+similar to the existing [prometheus
+source](https://vector.dev/docs/reference/sources/prometheus/) that will scrape
+the Apache HTTP Server stats endpoint (provided by
+[`mod_status`](https://httpd.apache.org/docs/2.4/mod/mod_status.html)) on an
+interval and publish metrics to the defined pipeline.
+
+## Scope
+
+This RFC will cover:
+
+- A new source for Apache metrics
+
+This RFC will not cover:
+
+- Generating metrics from Apache logs
+
+## Motivation
+
+Users running httpd want to collect, transform, and forward metrics to better
+observe how their webservers are performing.
+
+## Internal Proposal
+
+I expect to largely copy the existing [prometheus
+source](https://github.com/timberio/vector/blob/61e806d01d4cc6d2a527b52aa9388d4547f1ebc2/src/sources/prometheus/mod.rs)
+and modify it to parse the output of the httpd status page which looks like:
+
+```text
+localhost
+ServerVersion: Apache/2.4.46 (Unix)
+ServerMPM: event
+Server Built: Aug  5 2020 23:20:17
+CurrentTime: Friday, 21-Aug-2020 18:41:34 UTC
+RestartTime: Friday, 21-Aug-2020 18:41:08 UTC
+ParentServerConfigGeneration: 1
+ParentServerMPMGeneration: 0
+ServerUptimeSeconds: 26
+ServerUptime: 26 seconds
+Load1: 0.00
+Load5: 0.03
+Load15: 0.03
+Total Accesses: 30
+Total kBytes: 217
+Total Duration: 11
+CPUUser: .2
+CPUSystem: .02
+CPUChildrenUser: 0
+CPUChildrenSystem: 0
+CPULoad: .846154
+Uptime: 26
+ReqPerSec: 1.15385
+BytesPerSec: 8546.46
+BytesPerReq: 7406.93
+DurationPerReq: .366667
+BusyWorkers: 1
+IdleWorkers: 74
+Processes: 3
+Stopping: 0
+BusyWorkers: 1
+IdleWorkers: 74
+ConnsTotal: 1
+ConnsAsyncWriting: 0
+ConnsAsyncKeepAlive: 0
+ConnsAsyncClosing: 0
+Scoreboard: ________________________________________________________W__________________.....................................................................................................................................................................................................................................................................................................................................
+```
+
+I'll use this to generate the following metrics:
+
+- `apache.uptime_seconds` (counter)
+- `apache.total_accesses` (counter; extended)
+- `apache.total_kilobytes` (counter; extended)
```

Comment:
I prefer 1. but think we should provide tools that allow folks to name/label metrics to a schema. Normalizing to a convention requires the convention to be widely used. You could perhaps have a transform that converts metrics into prom naming rules?

Reply:
I feel like I could be swayed either way at this point, but I lean towards keeping it close to the source data and being otherwise unopinionated and let people map them. I personally like the promotheus naming rules, but they certainly haven't been universally adopted.

As another datapoint, OpenTelemetry is discussing their conventions over here: https://github.com/open-telemetry/opentelemetry-specification/issues/651